state.go

```diff
@@ -19,6 +19,10 @@ type State struct {
 	remoteCertificate     *x509.Certificate
 
 	isClient bool
+
+	serverKeyExchange    *handshakeMessageServerKeyExchange
```

Comment:
Have you ever read anything about what may/may be kept in memory?

`crypto/tls` itself only keeps the `masterSecret []byte` so I think we should push back on this

Reply:
So two problems that I ran into:
1) In order to calculate the `extendedMasterSecret` hash, you have to keep a binary "log" of the handshakes, until you calculate the hash
2) For the client, the last part of the handshake is sending out the `clientKeyExchange` so we have to wait for the `initalizeCipherSuite` to run at that point which then calculates the `extendedMasterSecret` and uses the serverKeyExchange  (specifically `publicKey`, `namedCurve`, `signature` and `hashAlgorithm` from it) to validate the key signature (if not disabled)

I'm happy to see if there are ways to more directly scrub it from memory but I am struggling to see how I (or another library?) could not keep it around

Reply:
Can we pull the `*handshakeMessageServerKeyExchange` from the `*handshakeCache`?

Reply:
Even the handshake cache would still be an unprotected memory region right?

We could consider introducing a dependency on https://godoc.org/github.com/awnumar/memguard, for creating a secure enclave where we store sensitive information. It's a little more extreme than just zeroing out a buffer but it goes to fairly great lengths to help you keep a memory area protected.

Reply:
@Sean-Der Yes, we could pull it from the handshake cache but I would have to unmarshal it again which would be a bit extra computation. I don't have a strong feeling either way

@daenney Yes, I think it's definitely worthwhile to have a discussion on how we could better protect memory in these cases. I opened up an issue here so we can have a fuller discussion on it: https://github.com/pion/dtls/issues/97

Reply:
My preference is double computation. Until we measure we don’t know how expensive something is.

But the more state we have the more difficult things become to keep in your head/and we have to guard and more nil checks (that’s just like my opinion though man)

Reply:
@daenney that’s super interesting I had no idea about memguard. 

I will play with it/understand the trade offs! I don’t have a security background/I am pretty unaware of what the community is doing :/